[FIX] Tooltip komponent

KenAJoh opened the pull request:
Third time's the charm?

Er implementert med FloatingUI og det fungerte greit. Kan migrere Popover etterhvert til nytt API

Radix sin implementasjon støttet ikke UU godt nok, så måtte lage den selv likevel. 95% av implementasjonen er lånt fra Material UI og Radix.

https://github.com/mui/material-ui/blob/v5.5.1/packages/mui-material/src/Tooltip/Tooltip.js
https://github.com/radix-ui/primitives/tree/main/packages/react/tooltip/src
https://github.com/radix-ui/primitives/blob/main/packages/react/portal/src/Portal.tsx


Forbedringer: 
- Tooltip flyter ikke over skjermbredde lengre, Forsvinner ikke på hover.

Hvorfor er vi så strenge med charLength, content: string og keys: string[] ?
- Tooltip skal ikke kunne inneholde avansert innhold, derfor ingen `ReactNode`
- Skal fungere som et "forbedret" `title`-element, derfor ønsker vi å oppmuntre til korte tekster.

KenAJoh committed:
css

KenAJoh committed:
react wip

KenAJoh committed:
forenklet tooltip

KenAJoh committed:
id-handling

KenAJoh committed:
Merge branch 'master' into new/tooltip-radix

KenAJoh committed:
yarn

KenAJoh committed:
fikset shadows og tokens

KenAJoh committed:
Update @navikt/core/css/tooltip.css

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/css/tooltip.css

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
oppdatert bruk av offset

KenAJoh committed:
padding

KenAJoh committed:
Merge branch 'master' into new/tooltip-radix

KenAJoh committed:
oppdatert yarn lock

KenAJoh committed:
fjernet key og justert shadow

KenAJoh committed:
stacket drop-shadow

KenAJoh committed:
oppdatert keys i tooltip

KenAJoh committed:
lint

KenAJoh committed:
uu-demo

KenAJoh committed:
Merge branch 'master' into new/tooltip-radix

KenAJoh committed:
yarn

KenAJoh committed:
fikset dobble opplesning av tekst skjermleser

KenAJoh committed:
Merge branch 'master' into new/tooltip-radix

KenAJoh committed:
startet ny tooltip implementasjon

KenAJoh committed:
styling

KenAJoh committed:
la til keys

KenAJoh committed:
mindre padding for hide

KenAJoh committed:
oppdatert api

KenAJoh committed:
fikset margin fix for hover state

KenAJoh committed:
fungerende hover og fokus interaksjon

KenAJoh committed:
oppdatert aria

KenAJoh committed:
oppdatert animasjon

KenAJoh committed:
oppdatert styling

KenAJoh committed:
fjernet kommented kode

KenAJoh committed:
[FIX] Alert styling og ikoner (#1447)

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>
Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>
Co-authored-by: Ken <ken.aleksander@gmail.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

github-actions[bot] committed:
Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.6
 - @navikt/ds-react@0.17.9
 - @navikt/ds-react-internal@0.12.9

KenAJoh committed:
[FIX] Endringsforslag Pagination (#1457)

github-actions[bot] committed:
Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.7
 - @navikt/ds-react@0.17.10
 - @navikt/ds-react-internal@0.12.10

macbruker committed:
[FIX] Replace hard coded values with tokens (#1463)

* Replace hard coded values with tokens

* Update yarn.lock

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

github-actions[bot] committed:
Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.8
 - @navikt/ds-react@0.17.11
 - @navikt/ds-react-internal@0.12.11

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+}
+
+.navds-tooltip {
+  z-index: 2147483647;
```
Denne burde ligge som en token.

KenAJoh replied:
God ide! Fikset 👍

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+  border-radius: var(--navds-border-radius-medium);
+  padding: 2px var(--navds-spacing-2);
+  align-items: center;
+  filter: drop-shadow(0 1px 2px rgba(0, 0, 0, 0.05))
```
Kjedelig å ikke få brukt skygge-tokens her.

KenAJoh replied:
Ja, er samme problemet som popover der vi evt kan ha shadow på selve tooltip, men ikke arrow da.

andnorda replied:
Kanskje vi burde ha `filter: drop-shadow()`-varianter av alle skygge-tokens?

KenAJoh replied:
God ide, kan se på det etterhvert.

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+  background-color: var(--navds-global-color-gray-700);
+  color: var(--navds-semantic-color-text-inverted);
+  border-radius: 2px;
+  padding: 0 4px;
```
```suggestion
  padding: 0 var(--navds-spacing-1);
```

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+  font-family: var(--navds-font-family);
+  background-color: var(--navds-global-color-gray-700);
+  color: var(--navds-semantic-color-text-inverted);
+  border-radius: 2px;
```
```suggestion
  border-radius: var(--navds-border-radius-small);
```

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+}
+
+.navds-tooltip__key + .navds-tooltip__key {
+  margin-left: 0.125rem;
```
Jeg ser det er brukt `2px` noen steder, og `0.125rem`. Egentlig burde vi kanskje innlemme 2px og 6px i spacing-skalaen når vi bruker det så mye. Kan det være en ide å gjøre om skalaen til å at spacing-1 er 2px?

KenAJoh replied:
Hørte med designer og justere tooltip til å ikke bruke noen av 2px/6px variantene. Kan se om vi justerer tabs også etterhvert til å ikke bruke dem.

andnorda replied:
Lurt

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+  /**
+   * Element tooltip anchors to
+   */
+  children: React.ReactElement & React.RefAttributes<HTMLElement>;
```
Er det noen grunn til å ikke bare bruke `ReactNode` her?
```suggestion
  children: React.ReactNode;
```

KenAJoh replied:
Ja er ikke mulig å bruke flere children, eks 
```jsx
<Tooltip>
    <Child1/>
    <Child2/>
</Tooltip>
```

Er heller ikke lov med react-fragments
```jsx
<Tooltip>
    <>
        <Child1/>
    </>
</Tooltip>
```

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+                  ...(staticSide
+                    ? { [staticSide[1]]: _offset ? _offset : _arrow ? 10 : 2 }
+                    : ""),
+                  opacity: 1,
```
Denne trengs ikke

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+      right: ["left", "marginLeft"],
+      bottom: ["top", "marginTop"],
+      left: ["right", "marginRight"],
+    }[placement.split("-")[0]];
```
Jeg synes `staticSide[0 | 1]` er litt for kryptisk. Vil du se om du finner en mer leselig variant? Kanskje inline noe greier de to stedene det brukes?

KenAJoh replied:
Endte opp med å inline det, eks:
```
[{
  top: "marginBottom",
  right: "marginLeft",
  bottom: "marginTop",
  left: "marginRight",
}[placement]]: _offset ? _offset : _arrow ? 10 : 2,
```

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+                left: x ?? "",
+                visibility: referenceHidden ? "hidden" : "visible",
+              }}
+              data-side={placement.split("-")[0]}
```
Vi trenger vel ikke å splitte denne når vi bare tillater `"top" | "right" | "bottom" | "left"`.

KenAJoh replied:
👍

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+  margin-left: 0.125rem;
+}
+
+.navds-tooltip__key--inverted {
```
Kan vi gjøre noe sånt, og så fjerne `navds-tooltip__key--inverted`-klassen?
```suggestion
.navds-tooltip--inverted .navds-tooltip__key {
```

KenAJoh replied:
👍

andnorda commented on @navikt/core/css/tooltip.css:
```diff
@@ -0,0 +1,73 @@
+}
+
+.navds-tooltip__key + .navds-tooltip__key {
+  margin-left: 0.125rem;
```
Kan dette heller gjøres med `flex` og `gap` på parent?

KenAJoh replied:
👍 Fikset!

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+      <>
+        {cloneElement(children, {
+          ...children.props,
+          ...ariaProps,
```
Noe sånt?
```suggestion
    return (
      <>
        {cloneElement(children, {
          ...children.props,
          aria-describedby={open ?? isOpen ? cl(ariaId, children?.props["aria-describedby"]) : null}
```

KenAJoh replied:
Endte opp med denne:
```jsx
          "aria-describedby":
            open || isOpen
              ? cl(ariaId, children?.props["aria-describedby"])
              : children?.props["aria-describedby"]
              ? children?.props["aria-describedby"]
              : undefined,
```

andnorda replied:
Er det det samme som dette?
```
  "aria-describedby":
            open || isOpen
              ? cl(ariaId, children?.props["aria-describedby"])
              : children?.props["aria-describedby"]  ?? undefined,
```

KenAJoh replied:
Jepp, bruker denne da!

andnorda replied:
`?? undefined` på slutten er kanskje ikke noe vits, når jeg tenker meg om.

KenAJoh replied:
Sant det, fikser det

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+      id,
+      inverted = false,
+      keys,
+      id: _id,
```
Dobbel id
```suggestion
```

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+    const openTimerRef = useRef(0);
+    const isMouseDownRef = useRef(false);
+
+    const ariaId = useId(_id);
```
Dobbel id
```suggestion
    const ariaId = useId(id);
```

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,321 @@
+
+/**
+ * TODO: console.error for lange tekster, egen prop for å justere opp makslengde
+ */
```
Utdaterte kommentarer ass... 🤮
```suggestion
```

andnorda reviewed (commented):
Bra jobba! Jeg fant noen greier. :)

KenAJoh committed:
Update @navikt/core/css/tooltip.css

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/css/tooltip.css

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Pr fikser

KenAJoh committed:
max-index token

KenAJoh committed:
max-index token

KenAJoh committed:
bedre aria-describedby

KenAJoh committed:
justert padding og arrow

KenAJoh committed:
justerte spacing bruk

KenAJoh committed:
fjernet keys inverted klasse

KenAJoh committed:
bedre aria-describedby

KenAJoh committed:
placement uten split

KenAJoh committed:
inline staticSide

andnorda committed:
[fix] Label react node (#1466)

github-actions[bot] committed:
Lerna versionbump [ci skip]

 - @navikt/ds-react@0.17.12
 - @navikt/ds-react-internal@0.12.12

larseirikhansen committed:
[NEW] New/stepper (#1429)

* satt opp stepper react

* satt opp stepper css

* oppdatert react-kode

* førsteutkast styling

* ♿ aria-label added to icons

* Stepper spacing and last-child focus

* add aria-labelledby

* add aria-labelledby to sb

* fix spelling mistake

* Remove 'finished' step api-option

* refactor Stepper-children jsx

* create Stepper example for sb

* WIP: Stepper mdx with practical props examples

* intro mdx disabled state

* ds intro overridable tags example

* add flex gap and fix last-child margin

* step-line is now pseudo-element

* cleanup css

* last child no margin-bottom

* remove disabled state

* update storybook

* aria-current to active step

* styling and color fixes

* color change line step

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Update @navikt/core/react/src/stepper/Step.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* [FIX] Bedre width for search (#1455)

* Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.3

* [fix] Helptext uu fix (#1407)

* storybook no div in p fix

* move title inside svg

* extra helptext in story

* remove duplicate title

* remove aria-props and refactor

* remove unused style

* simplify styling some

* default help text content to body-short

* cleanup

* tabindex 0 på popover

* flyttet story

* fjernet tabindex

* Update @navikt/core/react/src/help-text/HelpText.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

Co-authored-by: Ken <ken.aleksander@gmail.com>
Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

* Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.4
 - @navikt/ds-react@0.17.4
 - @navikt/ds-react-internal@0.12.4

* [FIX] Oppdatert ikon-template, SSR-safe uuid (#1458)

* oppdatert ikon-template

* oppdatert pakkversjoner

* fikset pakkeversjoner

* oppdatert ikon-template

* fikset config

* Lerna versionbump [ci skip]

 - @navikt/ds-icons@0.8.5
 - @navikt/ds-react@0.17.5
 - @navikt/ds-react-internal@0.12.5

* [fix] Read more component (#1456)

* scrips yo

* unused import

* stylez

* aria-expanded

* add link to storybook

* oopsie

* read more focus style

* read more collapse animation

* Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.5
 - @navikt/ds-react@0.17.6
 - @navikt/ds-react-internal@0.12.6

* remove ariaLabelledBy as its in by default

* Update @navikt/core/css/stepper.css

* anchor tag is default overridable

* rm space

* default from button to anchor

* add y-axis click affordance to steps

* click affordance, flex-align-start and stretching h-line with long label

* Update @navikt/core/css/stepper.css

Co-authored-by: Ken <ken.aleksander@gmail.com>
Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

github-actions[bot] committed:
Lerna versionbump [ci skip]

 - @navikt/ds-css@0.16.9
 - @navikt/ds-react@0.17.13
 - @navikt/ds-react-internal@0.12.13

KenAJoh committed:
[FIX] fikset tabs-scroll når scrollleft=0

github-actions[bot] committed:
Lerna versionbump [ci skip]

 - @navikt/ds-react@0.17.14
 - @navikt/ds-react-internal@0.12.14

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,322 @@
+                          left: "marginRight",
+                        }[placement]]: _offset ? _offset : _arrow ? 10 : 2,
+                      }
+                    : ""),
```
```suggestion
                  [{
                    top: "marginBottom",
                    right: "marginLeft",
                    bottom: "marginTop",
                    left: "marginRight",
                  }[placement]]: _offset ? _offset : _arrow ? 10 : 2,
```

andnorda commented on @navikt/core/react/src/tooltip/Tooltip.tsx:
```diff
@@ -0,0 +1,322 @@
+                              left: "right",
+                            }[placement]]: "-3.5px",
+                          }
+                        : ""),
```
```suggestion

                      [{
                        top: "bottom",
                        right: "left",
                        bottom: "top",
                        left: "right",
                      }[placement]]: "-3.5px",
```

andnorda commented on @navikt/core/tokens/src/index.js:
```diff
@@ -43,6 +43,7 @@ module.exports = {
       modal: { value: 2000 },
       popover: { value: 1000 },
       focus: { value: 10 },
+      max: { value: 2147483647 },
```
Kanskje denne også burde ha et semantisk navn som de andre? Og er det et poeng at den skal være absolutt max?
```suggestion
      tooltip: { value: 3000 },
```

KenAJoh replied:
Oppdaterte til `navds-z-index-tooltip` 🚀  Tror absolutt max bare er en nice to have her siden man sikrer at tooltip alltid er øverst, men ser ingen problem med å bare bruke en token med eks 3000 her.

andnorda reviewed (dismissed):
Ser bra ut nå! ⬐

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
fikset rebase

KenAJoh committed:
fikset rebase

KenAJoh committed:
Update @navikt/core/react/src/tooltip/Tooltip.tsx

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
Merge branch 'master' into new/tooltip-custom-v2

KenAJoh committed:
Update @navikt/core/tokens/src/index.js

Co-authored-by: Andreas Nordahl <andnorda@gmail.com>

KenAJoh committed:
oppdatert token-bruk

andnorda reviewed (dismissed):
Ship it! 🛳

KenAJoh committed:
fjernet close på click

KenAJoh committed:
Merge branch 'master' into new/tooltip-custom-v2

KenAJoh committed:
justert tooltip styling

KenAJoh committed:
fikset leave/enter problem på tooltip

KenAJoh committed:
fjernet inverted
